lib/test.dart

```diff
@@ -32,27 +56,103 @@ Future<FileContext> fileContextForTest(String name, String sourceText) async {
+/// will enable the analyzer to resolve imports and symbols from other source
+/// files and dependencies.
 ///     test('MySuggestor', () async {
-///       var context = await fileContextForTest('foo.dart', 'library foo;');
+///       var pkg = await PackageContextForTest.fromPubspec('pkg', '''
+///     name: pkg
+///     version: 0.0.0
+///     environment:
+///       sdk: '>=3.0.0 <4.0.0'
+///     dependencies:
+///       meta: ^1.0.0
+///     ''');
+///       var context = await pkg.addFile('foo.dart', '''
+///     import 'package:meta/meta.dart';
+///     @visibleForTesting var foo = true;
+///     ''');
 ///       var suggestor = MySuggestor();
 ///       var expectedOutput = '...';
 ///       expectSuggestorGeneratesPatches(suggestor, context, expectedOutput);
 ///     });
-void expectSuggestorGeneratesPatches(
-    Suggestor suggestor, FileContext context, dynamic resultMatcher) {
-  expect(
-      suggestor(context)
-          .toList()
-          .then((patches) => applyPatches(context.sourceFile, patches)),
-      completion(resultMatcher));
+class PackageContextForTest {
+  final AnalysisContextCollection _collection;
+  final String _name;
+  final String _root;
+
+  /// Creates a temporary directory named [dirName] using the `test_descriptor`
+  /// package, installs dependencies with `dart pub get`, sets up an analysis
+  /// context for the package, and returns a [PackageContextForTest] wrapper
+  /// that allows you to add source files to the package and use them in tests.
+  ///
+  /// Throws an [ArgumentError] if it fails to install dependencies.
+  static Future<PackageContextForTest> fromPubspec(
+    String dirName,
+    String pubspecContents,
+  ) async {
+    await d.dir(dirName, [
+      d.file('pubspec.yaml', pubspecContents),
+    ]).create();
+
+    final root = p.canonicalize(d.path(dirName));
+    final pubGet =
+        Process.runSync('dart', ['pub', 'get'], workingDirectory: root);
+    if (pubGet.exitCode != 0) {
+      printOnFailure('''
+PROCESS: dart pub get
+WORKING DIR: $root
+STDOUT:
+${pubGet.stdout}
+STDERR:
+${pubGet.stderr}
+''');
+      throw ArgumentError('Failed to install dependencies from given pubspec');
+    }
+    final collection = AnalysisContextCollection(includedPaths: [root]);
+    return PackageContextForTest._(dirName, root, collection);
+  }
+
+  PackageContextForTest._(this._name, this._root, this._collection);
+
+  /// Creates a temporary file at the given [path] (relative to the root of this
+  /// package) with the given [sourceText] using the `test_descriptor` package
+  /// and returns a [FileContext] wrapper around it.
+  ///
+  /// The returned [FileContext] will use the analysis context for this whole
+  /// package rather than just this file, which enables testing of [Suggestor]s
+  /// that require the resolved AST.
+  ///
+  /// See [PackageContextForTest] for an example.
+  Future<FileContext> addFile(String path, String sourceText) async {
```

Comment:
In over_react_codemod, we ended up having to share contexts across tests, since it would take too long to spin up a new analysis context for each test.

With that setup, it's a bit annoying to have to come up with unique file names for each test case, so being able to auto-generate them is nice (and what we ended up doing [in over_react_codemod](https://github.com/Workiva/over_react_codemod/blob/bb73b48f65e0e4fa4ee53f21549a74f0f45351c7/test/resolved_file_context.dart#L188)).

Reply:
Yeah I was wondering about that.. I was about to type something out about how it's unfortunate that the `test_descriptor` package uses `addTearDown` by default since that effectively scopes everything it creates to a single test, but then I found this issue in that repo for that exact case:
https://github.com/dart-lang/test_descriptor/issues/11

And it turns out they [solved it](https://github.com/dart-lang/test/issues/712) by updating how `addTearDown` from the test package works so that it knows when is called from `setUpAll`, which is awesome! TIL! So that means consumers could create a `PackageContextForTest` in a `setUpAll` to share it across tests without any other changes here. With that in mind I agree, auto-generating a file name would be convenient. I had wondered about auto-generating a name for the directory as well, so I think I'll do that as well.

Reply:
Awesome, thanks for adding that!

Oh nice, I was wondering how if the sharing might not be possible test_descriptor; glad that works with `setUpAll`!